### PR TITLE
Feature/157878166 make right gradient legend dynamic

### PR DESCRIPTION
**Pull request by lingyun1010**

### CODE REVIEW

File: __test__/GeneExpressionTSnePlot.test.js [hunk 1]
```diff
@@ -160,4 +161,5 @@ describe(`GeneExpressionTSnePlot`, () => {
 
     expect(wrapper.find(MultiStopGradient).length).toBe(1)
   })
+  */
```
Comment (alfonsomunozpomer) [inline __test__/GeneExpressionTSnePlot.test.js]: Don’t comment out the test, just delete it and replace it for a new one if needed.

Also, if the `MultiStopGradient` is no longer used delete all the related files and the imports where it’s used.

File: src/GeneExpressionTSnePlot.js [hunk 2]
```diff
@@ -7,6 +7,14 @@ import AtlasAutocomplete from 'expression-atlas-autocomplete'
 import MultiStopGradient from './MultiStopGradient'
 
 import './util/MathRound'
+import Responsive from 'react-responsive';
```
Comment (alfonsomunozpomer) [inline src/GeneExpressionTSnePlot.js:9]: Cool! I think this is going to be useful for some of our Highcharts woes.

File: src/GeneExpressionTSnePlot.js [hunk 3]
```diff
@@ -47,39 +55,52 @@ const _colourize = (colourRanges, defaultColour = `blue`, alpha = 0.65) => {
...
-          return {
+    return (plotData) => plotData.series.map((aSeries) => {
+      // I can’t think of a better way to reconcile series.name being a string and highlightSeries being an array of
+      // numbers. For more flexibility we might think of having our series be identified by an arbitrary ID string
+
+      if (!highlightSeries.length || highlightSeries.map((hs) => String(hs)).includes(aSeries.name)) {
+        return {
+          name: aSeries.name,
+          data: aSeries.data.map((point) => {
+            if(point.expressionLevel>0){
```
Comment (alfonsomunozpomer) [inline src/GeneExpressionTSnePlot.js]: Always use whitespace around mathematical and boolean operators.

File: src/GeneExpressionTSnePlot.js [hunk 4]
```diff
@@ -47,39 +55,52 @@ const _colourize = (colourRanges, defaultColour = `blue`, alpha = 0.65) => {
...
 
 const GeneExpressionScatterPlot = (props) => {
   const {atlasUrl, suggesterEndpoint, geneId, onSelectGeneId, speciesName} = props       // Suggester
   const {height, plotData, expressionGradientColours, highlightClusters} = props  // Chart
   const {loading, resourcesUrl, errorMessage} = props                       // Overlay
-
+  const colourSchema = [`#d4e4fb`,`#95adde`,`#6077bf`,`#1151D1`,`#35419b`,`#0e0573`] // light blue to dark blue
+  const colourSchemaLength = colourSchema.length
+  
+  const plotDisable = plotData.max===null
```
Comment (alfonsomunozpomer) [inline src/GeneExpressionTSnePlot.js]: I suggest you to change this to `Boolean(plotData.max)`, so whoever’s reading the code has a clear picture of your intent.

File: src/GeneExpressionTSnePlot.js [hunk 5]
```diff
@@ -47,39 +55,52 @@ const _colourize = (colourRanges, defaultColour = `blue`, alpha = 0.65) => {
...
+  const dataScale = plotDisable ? 0:plotData.max.toFixed(0).toString().length // The digit before demical
```
Comment (alfonsomunozpomer) [inline src/GeneExpressionTSnePlot.js]: Indent the ternary operator like this:
```
const dataScale = plotDisable ?
  0 : 
  plotData.max.toFixed(0).toString().length // The digit before demical
```

File: src/GeneExpressionTSnePlot.js [hunk 6]
```diff
@@ -93,25 +114,51 @@ const GeneExpressionScatterPlot = (props) => {
       }
     },
     chart: {
-      height: height
+      height: plotDisable ?  height*0.95 : height,
```
Comment (alfonsomunozpomer) [inline src/GeneExpressionTSnePlot.js]: See comments above regarding the ternary operator and mathematical operators.

File: src/GeneExpressionTSnePlot.js [hunk 7]
```diff
@@ -93,25 +114,51 @@ const GeneExpressionScatterPlot = (props) => {
...
     },
     title: {
       text: `Gene expression`
     },
-    legend: {
-      enabled: false
-    }
+    colorAxis: plotDisable ? {} :
+    {
```
Comment (alfonsomunozpomer) [inline src/GeneExpressionTSnePlot.js:128]: Review this block and fix indentation/spacing.

File: src/GeneExpressionTSnePlot.js [hunk 8]
```diff
@@ -93,25 +114,51 @@ const GeneExpressionScatterPlot = (props) => {
...
+      stop: colourSchema.map((val,idx) => {
+        return idx <= (Math.min(dataScale,colourSchemaLength) -1) ? [(idx+1)/Math.min(dataScale,colourSchemaLength),val] : []
+      }),
+      minColor:`rgb(215, 255, 255)`,
+      maxColor: dataScale>colourSchemaLength ? colourSchema[colourSchemaLength-1] : colourSchema[dataScale-1],
+      marker: {
+         color: '#c4463a'    
+      }
+    },
+    legend: plotDisable ? {enabled: false} :
```
Comment (alfonsomunozpomer) [inline src/GeneExpressionTSnePlot.js]: Apply our code conventions regarding string literals (we only use backticks).

File: src/plotloader/HeatmapLegend.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
+(function(H){"object"===typeof module&&module.exports?module.exports=H:H(Highcharts)})(function (H) {
```
Comment (alfonsomunozpomer) [inline src/plotloader/HeatmapLegend.js]: Clean up this code using our style as mentioned above.

Add a comment to the top of the file describing what this does and why it’s applied as a function on the Highcharts object.

Usually we reserve names with an initial upper case character like `HeatmapLegend.js` to React components. I think a better name for this file would be `highchartsHeatmapLegendMixin.js` or `highchartsHeatmapLegendModule.js`.

File: src/plotloader/PlotLoader.js [hunk 10]
```diff
@@ -16,7 +16,7 @@ const PlotLoader = ({loading, series, errorMessage, highchartsConfig, resourcesU
       <ScatterPlot chartClassName={chartClassName}
                    series={series}
                    highchartsConfig={highchartsConfig}
-                   children={children}
```
Comment (alfonsomunozpomer) [inline src/plotloader/PlotLoader.js:19]: I was never proud of adding `children` this way. Thank you! 😄

File: src/plotloader/ScatterPlot.js [hunk 11]
```diff
@@ -3,39 +3,77 @@ import PropTypes from 'prop-types'
...
+import HighchartsMap from 'highcharts/modules/map'
 import deepmerge from 'deepmerge'
 
+import HeatmapLegend from './HeatmapLegend'
 import SeriesPropTypes from './SeriesPropTypes'
+import Yaxispanning from './Yaxispanning'
+
 
-const Highcharts = ReactHighcharts.Highcharts
+const Highcharts = ReactHighcharts.Highcharts;
```
Comment (alfonsomunozpomer) [inline src/plotloader/ScatterPlot.js]: Remove semicolons here and everywhere else.

File: src/plotloader/ScatterPlot.js [hunk 12]
```diff
@@ -3,39 +3,77 @@ import PropTypes from 'prop-types'
...
+
+  await HighchartsExporting(Highcharts);
+  await HighchartsBoost(Highcharts);
+  await HighchartsMap(Highcharts);
+  await HighchartsHeatmap(Highcharts);
+  await HeatmapLegend(Highcharts);
+  await Yaxispanning(Highcharts);
+}
+
+addModules();
```
Comment (alfonsomunozpomer) [inline src/plotloader/ScatterPlot.js]: Why are these modules wrapped in an asynchronous function?

File: src/plotloader/ScatterPlot.js [hunk 13]
```diff
@@ -3,39 +3,77 @@ import PropTypes from 'prop-types'
...
+    height: `100%`,
+    panning: true,
+    spacingTop: 50,
+  },
+
+  mapNavigation: {
+    enabled: true,
+    enableMouseWheelZoom: false,
+    buttonOptions: {
+      theme: {
```
Comment (alfonsomunozpomer) [inline src/plotloader/ScatterPlot.js]: Again, use backticks instead of single quotes.

File: src/plotloader/ScatterPlot.js [hunk 14]
```diff
@@ -76,34 +116,36 @@ const ScatterPlot = (props) => {
         plotOptions: {
           series: {
             marker: {
-              radius: numPoints < 5000 ? 4 : 0.2
+              radius:  3
```
Comment (alfonsomunozpomer) [inline src/plotloader/ScatterPlot.js]: Remove extra single space.

File: src/plotloader/ScatterPlot.js [hunk 15]
```diff
@@ -76,34 +116,36 @@ const ScatterPlot = (props) => {
...
             }
           }
         }
       },
       { series: series },
-      highchartsConfig
+      highchartsConfig,
+      {
+        legend: {
+          symbolWidth: legendWidth
```
Comment (alfonsomunozpomer) [inline src/plotloader/ScatterPlot.js]: See above...

File: src/plotloader/Yaxispanning.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
+/**
```
Comment (alfonsomunozpomer) [inline src/plotloader/Yaxispanning.js]: Apply our styles to this file. Rename it similarly to the suggestion for the heatmap legend module. Also, use camelCase for the name.

Also, rewrite the comments as the purpose of this module differs to what’s stated in the preamble. You can keep an attribution of the original author if you think it should be credited.

**Review by alfonsomunozpomer — CHANGES_REQUESTED** [review]
Functionally the code is good but you need to take extra care about code guidelines, punctuation, names of files and variables, etc. In that sense it’s a bit rough but you’ll improve with practice. When you code in a team these things become all the more important since other people will read your code and you need to transmit your ideas through code.

Please go through my comments and add all the suggested modifications.

**Comment by monicajianu** [comment]
I mentioned this in the [first iteration of this PR](https://github.com/ebi-gene-expression-group/scxa-experiment-page-tsne-plot-view/pull/11), but I don't think the +/- zooming buttons have a good enough performance to be released in production, which is a bit of a shame. @alfonsomunozpomer, what do you think?

**Comment by alfonsomunozpomer** [comment]
I also think that the +/- zoom buttons need to be reworked or improved. Besides what you point out, there’s no “Reset zoom” button, which is a bit annoying. Once the code is fixed we’ll discuss what to do with the zoom feature.

File: src/plotloader/ScatterPlot.js (hunk 12 above)
Comment (lingyun1010) [reply src/plotloader/ScatterPlot.js]: I added a comment in [Pivotal](https://www.pivotaltracker.com/story/show/157878166/comments/193178664) about this issue before. Please have a look into it.

**Review by alfonsomunozpomer — CHANGES_REQUESTED** [review]
The Y-axis panning module isn’t loaded correctly. As the spelling’s changed you need to change it in the `async` block too.

For the time being, could you comment out the modules for the +/- zoom and the Y-axis panning, please? They’re very helpful so don’t delete them. I’ll be adding a story in the next sprint to combine both zooming behaviours.

**Review by alfonsomunozpomer — APPROVED** [review]

**Pull request merged**